src/net_processing.cpp

```diff
@@ -6264,8 +6264,29 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                             }
                         }
 
-                        if (fanout || !m_txreconciliation->AddToSet(pto->GetId(), wtxid).m_succeeded) {
+                        if (fanout) {
                             vInv.push_back(inv);
+                        } else {
+                            // If the transactions fails to get into the set, we fanout
+                            auto result = m_txreconciliation->AddToSet(pto->GetId(), wtxid);
+                            if (!result.m_succeeded) {
+                                vInv.push_back(inv);
+                                // If the transaction fails because it collides with an existing one,
+                                // we also remove and fanout the conflict and all its descendants.
+                                // This is because our peer may have added the conflicting tranaction
+                                // to its set, in which reconciliation of these two would fail
+                                auto collision = result.m_conflict;
+                                if (collision.has_value()) {
+                                    Assume(peer->m_wtxid_relay);
+                                    CTxMemPool::setEntries descendants;
+                                    m_mempool.CalculateDescendants(m_mempool.get_iter_from_wtxid(collision.value()), descendants);
+                                    for (const auto &txit: descendants) {
+                                        auto wtxid = txit->GetTx().GetWitnessHash();
+                                        m_txreconciliation->TryRemovingFromSet(pto->GetId(), wtxid);
+                                        vInv.emplace_back(MSG_WTX, wtxid);
```

Comment:
That's a good point. It may be the case that the selected peer has a parent -> child dependency but was not picked as one of the `fanout_with_ancestors` peers, the parent and the child may have ended up in different sets.

We should check the return of `TryRemovingFromSet` before adding it to `vInv`

Reply:
I just realized this is also the case for the ancestors. Went relaying with ancestors, we should also check that the ancestor we are trying to relay is part of the reconciliation set (i.e. that it has not been previously flooded)